File: src/Server/base/Caducar.js
const express = require('express'); 
const multer = require('multer');
const fs = require('fs');
const path = require('path');
const db = require('./connection');
const router = express.Router();

router.get("/:id", (req, res) => {
  console.log("Llamando a caducar");

  const query1 = `
    SELECT 
      sd.Id_Stock_Detalle AS id,
      ca.Alimento AS Nombre,
      sd.Cantidad AS Cantidad,
      sd.Fecha_Caducidad AS Fecha,
      ca.Id_Usuario_Alta
    FROM stock_detalle sd
    LEFT JOIN cat_alimento ca ON sd.Id_Alimento = ca.Id_Alimento
    WHERE ca.Es_Perecedero = 1 
      AND sd.Activo = 1 
      AND ca.Id_Usuario_Alta = ?
    ORDER BY sd.Fecha_Caducidad ASC 
    LIMIT 10;
  `;

  const id = req.params.id; // Se obtiene correctamente el ID desde los parámetros

  db.query(query1, [id], (err, result1) => {
    if (err) {
      console.error("Error:", err);  // Usa console.error para errores
      return res.status(500).json({ error: "Error en el servidor" });
    }
    
    if (result1.length === 0) {
      return res.status(404).json({ message: "No se encontraron alimentos por caducar" });
    }

    console.log("Enviando alimentos por caducar:", result1);
    
    // Enviar la respuesta correctamente
    return res.json({ porcaducar: result1 });
  });
});



router.put("/", (req, res) => {
  const { ids } = req.body; // Obtener los IDs desde el cuerpo de la solicitud

  // Validar que se hayan enviado los datos necesarios
  if (!ids || !Array.isArray(ids) || ids.length === 0) {
    return res.status(400).send("Faltan los IDs o no son válidos.");
  }

  // Registrar resultados de las actualizaciones
  const resultados = { actualizados: [], errores: [] };

  // Función para procesar cada ID
  const procesarId = (id, callback) => {
    const query = `UPDATE stock_detalle SET Activo = 0, Cantidad = 0 WHERE Id_Stock_Detalle = ?`;

    db.query(query, [id], (err, result) => {
      if (err) {
        console.error(`Error al actualizar el ID ${id}:`, err);
        resultados.errores.push({ id, error: err.message });
      } else {
        console.log(`ID ${id} actualizado correctamente.`);
        resultados.actualizados.push(id);
      }
      callback(); // Llamar al callback para continuar
    });
  };

  // Procesar los IDs uno por uno
  let index = 0;
  const procesarSiguiente = () => {
    if (index < ids.length) {
      const id = ids[index];
      index++;
      procesarId(id, procesarSiguiente); // Procesar el siguiente ID
    } else {
      // Finalizar cuando se procesen todos los IDs
      res.json({
        message: "Proceso de actualización finalizado.",
        resultados,
      });
    }
  };

  // Iniciar el procesamiento
  procesarSiguiente();
});



module.exports = router;

File: src/serverP/base/Caducar.py
from fastapi import APIRouter, HTTPException
from .connection import get_db_connection
from pydantic import BaseModel
from typing import List

caducar_bp = APIRouter()

class IdsRequest(BaseModel):
    ids: List[int]

@caducar_bp.get("")
def obtener_por_caducar():
    connection = get_db_connection()
    cursor = connection.cursor(dictionary=True)

    query = """
        SELECT 
            sd.Id_Stock_Detalle AS id,
            ca.Alimento AS Nombre,
            sd.Cantidad AS Cantidad,
            sd.Fecha_Caducidad AS Fecha,
            ca.Id_Usuario_Alta
        FROM stock_detalle sd
        LEFT JOIN cat_alimento ca ON sd.Id_Alimento = ca.Id_Alimento
        WHERE ca.Es_Perecedero = 1 and sd.Activo = 1
        ORDER BY sd.Fecha_Caducidad ASC 
        LIMIT 10;
    """
    cursor.execute(query)
    alimentos = cursor.fetchall()

    cursor.close()
    connection.close()

    if not alimentos:
        raise HTTPException(status_code=404, detail="Alimento no encontrado")

    return {"porcaducar": alimentos}

@caducar_bp.put("/")
def actualizar_alimentos(request: IdsRequest):
    if not request.ids:
        raise HTTPException(status_code=400, detail="Faltan los IDs o no son válidos.")

    connection = get_db_connection()
    cursor = connection.cursor()

    resultados = {"actualizados": [], "errores": []}

    for id in request.ids:
        query = "UPDATE stock_detalle SET Activo = 0, Cantidad = 0 WHERE Id_Stock_Detalle = %s"
        try:
            cursor.execute(query, (id,))
            connection.commit()
            resultados["actualizados"].append(id)
        except Exception as e:
            connection.rollback()
            resultados["errores"].append({"id": id, "error": str(e)})

    cursor.close()
    connection.close()

    return {"message": "Proceso de actualización finalizado.", "resultados": resultados}
